feat: add import order rules

ritazoliveira opened the pull request:
In light of this [RFC](https://github.com/moxystudio/rfcs-oss/blob/master/rfcs/0004-eslint-config-import.md) two rules were added to our config: 

- `import/order`: displays an error when imports are not in the following order: `builtin`, `external`, `parent`, `sibling`, `index` and groups are not separated with a new line.

- `import/first`: displays an error when imports are made in the module's body.

As for `import/first`, since we already happen to have `import/order`, which already orders the module's import, this could be turned off in certain cases but maybe it's best to ensure all imports are made at the top of the file.

ritazoliveira committed:
feat: add import order rules

BREAKING CHANGE: added rules to order import

satazor commented on packages/eslint-config-base/lib/modules/esm.js:
```diff
@@ -9,5 +9,12 @@ module.exports = {
+        // Order imports
+        'import/order': ['error', {
+            groups: ['builtin', 'external', 'parent', 'sibling', 'index'],
+            'newlines-between': 'always',
```
I would either disable the new line or actually forbid it. Having new lines between groups is strange.

ritazoliveira replied:
Since it was suggested in the RFC I implemented it that way, but we can remove it for sure, I'm also not the biggest fan of forcing the newline after each group. @zebateira what do you think?

zebateira replied:
Yeah that's fine, we can remove it. I think it brings structure to the import section but I'm fine if this is not enforced. It's very subjective as well.

satazor commented:
@ritazoliveira I would also consider enabling https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/newline-after-import.md.

zebateira commented on __snapshots__/index.test.js.snap:
```diff
@@ -1318,22 +1318,88 @@ exports[`should pass on packages/eslint-config-base/test/samples/es2015-esm/cont
     "column": 1,
-    "line": 5,
+    "line": 3,
+    "rule": "lines-around-directive",
```
did we change this rule? https://eslint.org/docs/rules/lines-around-directive
It seems to be deprecated in favor of https://eslint.org/docs/rules/padding-line-between-statements

ritazoliveira replied:
We are using both actually, should we change this?

satazor replied:
We may remove lines-around-directive, but be sure it doesn’t cause troubles by checking snapshots.

satazor replied:
Actually they are different, we must keep both.

satazor replied:
See my comment below.

ritazoliveira commented:
Removed the mandatory new line after each import group and also added the [`newline-after-import`](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/newline-after-import.md.) rule.

satazor commented on packages/eslint-config-base/test/samples/es2015-esm/contrived.bad.js:
```diff
@@ -1,11 +1,25 @@
 // Made up sample.. not genuine
-
 // Test unnecessary `use strict`
-
```
Add this new line back to remove the `lines-around-directive` error from the snapshot.

ritazoliveira replied:
Fixed!

satazor commented:
This is a breaking change. @ritazoliveira next time don’t forget to sinal it in the commit message or body.

ritazoliveira commented:
Fixed the commit to include `BREAKING CHANGE` in the body.

satazor commented on __snapshots__/index.test.js.snap:
```diff
@@ -2757,6 +2805,12 @@ Array [
+  Object {
+    "column": 1,
+    "line": 5,
+    "rule": "import/newline-after-import",
```
Can you fix this error by fixing `packages/eslint-config-vue/test/rules/vue-no-shared-component-data.vue`?

satazor commented on __snapshots__/index.test.js.snap:
```diff
@@ -1318,22 +1318,70 @@ exports[`should pass on packages/eslint-config-base/test/samples/es2015-esm/cont
+  Object {
+    "column": 8,
+    "line": 8,
+    "rule": "no-unused-vars",
```
This shouldn't happen. Can you add a something with the `index` variable to suppress this error?

satazor commented on __snapshots__/index.test.js.snap:
```diff
@@ -1318,22 +1318,70 @@ exports[`should pass on packages/eslint-config-base/test/samples/es2015-esm/cont
+  Object {
+    "column": 8,
+    "line": 24,
+    "rule": "no-unused-vars",
```
Same.

ritazoliveira committed:
fix: lines with errors in snapshot
